Fix #14844

Clyybber opened the pull request:
One test will fail because the current broken behaviour allowed it to (https://play.nim-lang.org/#ix=2qAU).
Should both fail or work? A manual `(var i = 0); (var i = 0)` currently fails.. (for more context see https://github.com/nim-lang/Nim/commit/bb532a697edad1bac60a87a7ff43956c9635973d#diff-705c21e2cbb7aca092a9882547994008)

Clyybber committed:
Fix #14844

Clyybber committed:
This test fails

timotheecour commented:
* could you add tests cases, at least the ones in https://github.com/nim-lang/Nim/issues/14844#issuecomment-651295359 ? (ideally turning them into doAssert tests)

* likewise with top-level vs proc scope tests mentioned there

> Should both fail or work? 

given that all these fail:
```nim
when defined case6:
  template bar*() =
    var r = @[(var i = @[""]; i), (var i = @[""]; i)]
    echo r
  bar()

when defined case7:
  var r = @[(var i = @[""]; i), (var i = @[""]; i)]
  echo r

when defined case8:
  proc main=
    var r = @[(var i = @[""]; i), (var i = @[""]; i)]
    echo r
  main()
```

it's better to be consistent; as well as to be consistent between MCS and regular call syntax, so the logical thing to do is make both fail; if absolutely needed add a `-d:nimHasWorkaround14863` for a transition time. 

Note that I'm fixing `applyIt` here to avoid the double evaluation bug: https://github.com/nim-lang/Nim/pull/14869

Clyybber committed:
Cleanup and add test

Clyybber committed:
WIP

Clyybber committed:
Small cleanup

Clyybber committed:
Small cleanup

Clyybber committed:
WIP

timotheecour commented:
can `nkEarlySemArg` also be used to solve the long standing issue of const folding happening too early, eg:
* https://github.com/nim-lang/Nim/issues/9301
* https://github.com/nim-lang/Nim/issues/6293
* https://github.com/nim-lang/Nim/issues/11789
?

Clyybber committed:
Fix

Araq commented on compiler/ast.nim:
```diff
@@ -225,6 +225,7 @@ type
     nkBreakState,         # special break statement for easier code generation
     nkFuncDef,            # a func
     nkTupleConstr         # a tuple constructor
+    nkEarlySemArg         # When an argument was semmed early, but we might need the untyped AST later, the semmed AST is the first, the unsemmed the second son
```
Simply remembering both variants can create a compiler that uses an expontential amount of memory. (Nest these calls to get 2x2x2x2x... memory consumption.) There should only be a single AST that works with all the constructs. My "semtityped" AST idea.

Clyybber replied:
No worries, I have an unpushed fix for that.

Araq replied:
*Now* I worry...

Clyybber replied:
We just discard our reference to the unsemmed/semmed AST when choosing which one we need after sigmatch. https://github.com/nim-lang/Nim/pull/14863/commits/fb0e697bbdf20a4087dfd463ab018d2fbb1f683e

Araq commented:
This is very dangerous terrain.

Clyybber committed:
Fix

Clyybber committed:
Update test

Clyybber committed:
Fix wrong test.

It was testing that the behaviour is inconsistent
with what the code would look like written out manually :D

Clyybber committed:
Preliminary fixing

Clyybber committed:
Add expected output to tmethodcall

Clyybber committed:
Unref unused AST

Clyybber committed:
Small cleanup

Araq commented:
Too convoluted. Instead sem'checking should understand nkHiddenAddr, nkHiddenDeref and we need a `nkHiddenArrayConstr` node to make sem'checking idempotent.

Clyybber commented:
That isn't related. This is about preventing typed AST ending up in untyped arguments.

Araq commented:
And why is that a problem? "untyped" must go away, we need a solid typed AST so that macros can do type-based analysis. We don't need "I need untyped because the AST is then simpler", that's a symptom of the underlying issues.
